boa_engine/src/builtins/weak/weak_ref.rs

```diff
@@ -0,0 +1,166 @@
+            WellKnownSymbols::to_string_tag(),
+            "WeakRef",
+            Attribute::CONFIGURABLE,
+        )
+        .method(WeakRef::deref, "deref", 0)
+        .build()
+        .conv::<JsValue>()
+        .pipe(Some)
+    }
+}
+
+impl WeakRef {
+    /// The amount of arguments the `WeakRef` constructor takes.
+    pub(crate) const LENGTH: usize = 1;
+
+    /// Constructor [`WeakRef ( target )`][cons]
+    ///
+    /// [cons]: https://tc39.es/ecma262/#sec-weak-ref-target
+    pub(crate) fn constructor(
+        new_target: &JsValue,
+        args: &[JsValue],
+        context: &mut Context,
+    ) -> JsResult<JsValue> {
+        // If NewTarget is undefined, throw a TypeError exception.
+        if new_target.is_undefined() {
+            return Err(JsNativeError::typ()
+                .with_message("WeakRef: cannot call constructor without `new`")
+                .into());
+        }
+
+        // 2. If target is not an Object, throw a TypeError exception.
+        let target = args.get(0).and_then(JsValue::as_object).ok_or_else(|| {
+            JsNativeError::typ().with_message(format!(
+                "WeakRef: expected target argument of type `object`, got target of type `{}`",
+                args.get_or_undefined(0).type_of()
+            ))
+        })?;
+
+        // 3. Let weakRef be ? OrdinaryCreateFromConstructor(NewTarget, "%WeakRef.prototype%", « [[WeakRefTarget]] »).
+        // 5. Set weakRef.[[WeakRefTarget]] to target.
+        let weak_ref = JsObject::from_proto_and_data(
+            get_prototype_from_constructor(new_target, StandardConstructors::weak_ref, context)?,
+            ObjectData::weak_ref(WeakGc::new(target.inner())),
+        );
+
+        // 4. Perform AddToKeptObjects(target).
+        context.kept_alive.push(target.clone());
+
+        // 6. Return weakRef.
+        Ok(weak_ref.into())
+    }
+
+    /// Method [`WeakRef.prototype.deref ( )`][spec].
+    ///
+    /// If the referenced object hasn't been collected, this method promotes a `WeakRef` into a
+    /// proper [`JsObject`], or returns `undefined` otherwise.
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-weak-ref.prototype.deref
+    pub(crate) fn deref(this: &JsValue, _: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
+        // 1. Let weakRef be the this value.
+        // 2. Perform ? RequireInternalSlot(weakRef, [[WeakRefTarget]]).
+        let weak_ref = this.as_object().map(JsObject::borrow).ok_or_else(|| {
+            JsNativeError::typ().with_message(format!(
+                "WeakRef.prototype.deref: expected `this` to be an `object`, got value of type `{}`",
+                this.type_of()
+            ))
+        })?;
+
+        let weak_ref = weak_ref.as_weak_ref().ok_or_else(|| {
+            JsNativeError::typ()
+                .with_message("WeakRef.prototype.deref: expected `this` to be a `WeakRef` object")
+        })?;
+
+        // 3. Return WeakRefDeref(weakRef).
+
+        // `WeakRefDeref`
```

Comment:
I'm OK with this approach, but why does the spec move it out of the definition? Is it used elsewhere? If so, maybe we should use a macro, as we do with promises.

Reply:
The [spec](https://tc39.es/ecma262/multipage/managing-memory.html#sec-weakrefderef) says:

>This abstract operation is defined separately from WeakRef.prototype.deref strictly to make it possible to succinctly define liveness.

And the [Liveness](https://tc39.es/ecma262/multipage/executable-code-and-execution-contexts.html#sec-liveness) section of the spec uses `WeakRefDeref` as:

> For some set of objects S, a hypothetical WeakRef-oblivious execution with respect to S is an execution whereby the abstract operation [WeakRefDeref](https://tc39.es/ecma262/multipage/managing-memory.html#sec-weakrefderef) of a [WeakRef](https://tc39.es/ecma262/multipage/managing-memory.html#sec-weak-ref-constructor) whose referent is an element of S always returns undefined.
>
> At any point during evaluation, a set of objects S is considered live if either of the following conditions is met:
>
> - Any element in S is included in any [agent](https://tc39.es/ecma262/multipage/executable-code-and-execution-contexts.html#agent)'s [[KeptAlive]] [List](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#sec-list-and-record-specification-type).
> - There exists a valid future hypothetical WeakRef-oblivious execution with respect to S that observes the Object value of any object in S.

In short, it's just moved out by the spec's requirements.